src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,197 @@
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false
+        );
+
+        if ($this->buffer === false) {
+            $this->buffer = '';
+        }
+    }
+
+    protected function iterateBuffer()
+    {
+        if (strlen($this->buffer) <= 1) {
+            return false;
+        }
+
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false && strlen($this->buffer) > 1024) {
+                $this->emit('error', [
+                    new Exception('Chunk length header longer then 1024 bytes'),
+                ]);
+                $this->close();
+                return false;
+            }
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
+            $this->nextChunkIsLength = false;
+            $lengthChunk = substr($this->buffer, 0, $crlfPosition);
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk) = explode(';', $lengthChunk, 2);
+            }
+            if (dechex(hexdec($lengthChunk)) !== $lengthChunk) {
+                $this->emit('error', [
+                    new Exception('Unable to validate "' . $lengthChunk . '" as chunk length header'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->remainingLength = hexdec($lengthChunk);
+            $this->buffer = substr($this->buffer, $crlfPosition + 2);
+            return true;
+        }
+
+        if ($this->remainingLength > 0) {
+            $chunkLength = $this->getChunkLength();
+            if ($chunkLength === 0) {
+                return true;
+            }
+            $this->emit('data', array(
+                substr($this->buffer, 0, $chunkLength),
+                $this
+            ));
+            $this->remainingLength -= $chunkLength;
+            $this->buffer = substr($this->buffer, $chunkLength);
+            return true;
+        }
+
+        $this->nextChunkIsLength = true;
+        $this->buffer = substr($this->buffer, 2);
+
+        if (substr($this->buffer, 0, 5) === "0\r\n\r\n") {
+            $this->stream->removeListener('data', array($this, 'handleData'));
+            $this->stream->removeListener('end',  array($this, 'handleEnd'));
+            $this->emit('end');
+            return false;
```

Comment:
Something like this should do it:

```
$this->emit('end');
$this->close()`;
```

Yes, this will close the underlying stream, but IMO this is not something we should worry about right now. Connection reuse will require more effort anyway.

(Also, RFC 2616 is deprecated, you should consult https://tools.ietf.org/html/rfc7230#section-4.1.2 and in particular the following section instead)